### DISCUSSION TITLE
How to send to a "thank you" page after a form submit?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#37860] [Help] [answered]
I have this code here:

```
<Link href="/thankyou">
              <input
                type="submit"
                name="submit"
                value="submit"
                className="btn btn-primary"
                onClick={(e) => {
                  handleSubmit(e);
                }}
              />
            </Link>
```

^^^ it submits the form OK, but it doesn't send you to the "thank you" page..  would anybody know how to fix this ?

**Comment by Elindo586** [comment]
So... I am sending emails just fine from my localhot3000.... but not from my live website yet...  there are no errors on the npm run build, so I suppose the new site is uploaded, but not showing changes yet?

Anyway.... from the live website maybe with the old code, I can get to the "thank you page", but after changing the form there is something in the code that won't allow the thank you page come up.

**Comment by icyJoseph** [comment, marked as answer]
I'd do this in another way, do not wrap the input with a Link, and instead in your handler:

```jsx
  const router = useRouter(); // import `useRouter` from next/router

  const handleSubmit = (e) => {
    e.preventDefault();
    console.log("Sending");

    const data = {
      firstName,
      company,
      email,
      country,
      notes,
    };

    fetch("/api/contact", {
      method: "POST",
      headers: {
        Accept: "application/json, text/plain, */*",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    }).then((res) => {
      console.log("Response received");
      if (res.status === 200) {
         router.replace("/thank-you");
      }
    }).catch((err) => {/* handle failure */});
  };
```

**Reply by jelmd** [reply]
Why not simply `window.location.href = '/thank-you'` ? Why all the router bloat?

**Reply by icyJoseph** [reply]
That'd reload the page, but it is a fair observation. If what we want is a `SPA-transition` then we should use the router "bloat" as you call it, otherwise, there's other methods at hand.

We could also observe that we perhaps don't even need to prevent the default, and could perfectly just allow the page to navigate to an endpoint where we would collect and process the form data, and then, from the server, redirect the user to a given page.

- https://nextjs.org/docs/api-routes/response-helpers#redirects-to-a-specified-path-or-url

However, those would unload the page, which in some cases breaks the user experience, or work the user might have already done. It might also penalise the user with heavier data payloads in situations where proper caching strategies are not used.

**Comment by Elindo586** [comment]
Actually... . I was gonna use the handler, but apparently I didn't....    I did use the next/router with the useSatete.... and ended up somehow doing this and it is working...

```
 <input
          type="submit"
          name="submit"
          value="submit"
          className="btn btn-primary"
          onClick={(e) => {
            handleSubmit(e);
            (r) => {
              setRoute(r.target.value);
            };
          }}
        />

```

**Reply by Mustaqgithub** [reply]
javascript

**Comment by Elindo586** [comment]
I think... this is "when your code works, but you don't know why" type of cases...

```
const handleSubmit = (e) => {
    e.preventDefault();

    console.log("Sending");

    let data = {
      firstName,
      company,
      email,
      country,
      notes,
    };

    fetch("/api/contact", {
      method: "POST",
      headers: {
        Accept: "application/json, text/plain, */*",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    }).then((res) => {
      console.log("Response received");
      if (res.status === 200) {
        console.log("Response succeeded!");
        setSubmitted(true);
        setFirstName("");
        setCompany("");
        setEmail("");
        setCountry("");
        setNotes("");
      }
    });
    router.push("/thankyou");
  };
```

**Comment by icyJoseph** [comment]
A couple of things. `fetch` is asynchronous, which means that the `.then` callback runs later on time, exactly when the request resolves.

The way you have it there, `router.push` happens when fetch is queued, but not resolved. This is why I had placed the `router.push` inside the `then` callback, because we want to move to `thankyou` when the request returns with an ok answer.

Also, `route.replace` is a better choice here because it means that if the user click the back button on the browser, they'll not return to the contact form, which is in some cases, desired.

### Page unmounts

In this use case, when the request resolves there's no need to blank the states you are using.

Why? Because we route visitors to `/thankyou` which is a different page, and this means that the `/contact-us` page is unmounted, and all of the state inside it is discarded.

**Comment by Elindo586** [comment]
Thanks... I was working on this one by the end of the day and my head was already overloaded from figuring out the smtp email, among other things..... then I end up with extra code that isn't needed.
